src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/TypeMapLazyDictionary.cs

```diff
@@ -0,0 +1,448 @@
+            if (startingAssembly is null)
+            {
+                throw new InvalidOperationException(SR.InvalidOperation_TypeMapMissingEntryAssembly);
+            }
+
+            CallbackContext context;
+            ProcessAttributes(
+                new QCallAssembly(ref startingAssembly),
+                new QCallTypeHandle(ref groupType),
+                newExternalTypeEntry,
+                newProxyTypeEntry,
+                &context);
+
+            // If an exception was thrown during the processing of
+            // the attributes, rethrow it.
+            context.CreationException?.Throw();
+
+            return context;
+        }
+
+        public static IReadOnlyDictionary<string, Type> CreateExternalTypeMap(RuntimeType groupType)
+        {
+            unsafe
+            {
+                return CreateMaps(
+                    groupType,
+                    &NewExternalTypeEntry,
+                    null).ExternalTypeMap;
+            }
+        }
+
+        public static IReadOnlyDictionary<Type, Type> CreateProxyTypeMap(RuntimeType groupType)
+        {
+            unsafe
+            {
+                return CreateMaps(
+                    groupType,
+                    null,
+                    &NewProxyTypeEntry).ProxyTypeMap;
+            }
+        }
+
+        private abstract class LazyTypeLoadDictionary<TKey> : IReadOnlyDictionary<TKey, Type> where TKey : notnull
+        {
+            protected abstract bool TryGetOrLoadType(TKey key, [NotNullWhen(true)] out Type? type);
+
+            public Type this[TKey key]
+            {
+                get
+                {
+                    if (!TryGetOrLoadType(key, out Type? type))
+                    {
+                        ThrowHelper.ThrowKeyNotFoundException(key);
+                    }
+
+                    return type;
+                }
+            }
+
+            public bool TryGetValue(TKey key, [MaybeNullWhen(false)] out Type value) => TryGetOrLoadType(key, out value);
+
+            // Not supported to avoid exposing TypeMap entries in a manner that
+            // would violate invariants the Trimmer is attempting to enforce.
+            public IEnumerable<TKey> Keys => throw new NotSupportedException();
+            public IEnumerable<Type> Values => throw new NotSupportedException();
+            public int Count => throw new NotSupportedException();
+            public bool ContainsKey(TKey key) => throw new NotSupportedException();
+            public IEnumerator<KeyValuePair<TKey, Type>> GetEnumerator() => throw new NotSupportedException();
+            IEnumerator IEnumerable.GetEnumerator() => throw new NotSupportedException();
+        }
+
+        private unsafe struct TypeNameValue
+        {
+            public void* Utf8TypeName { get; init; }
+            public int Utf8TypeNameLen { get; init; }
+            public void* Utf8AssemblyNameFallback { get; init; }
```

Comment:
Instead of storing assembly name, it may be better to just store RuntimeAssembly here and then call `TypeNameResolver.GetTypeHelper` to resolve the type. `TypeNameResolver.GetTypeHelper` is the method that is used to [resolve type name in attributes](https://github.com/dotnet/runtime/blob/7cc32192cbb4894bfebbb09a903215993207a3c0/src/coreclr/vm/typeparse.cpp#L54-L59) and it will substitute the assembly name as needed. We won't need to worry about substituting the 
assembly name here or making sure that it is done exactly same way.

Reply:
Good idea.